dhis-2/dhis-test-integration/src/test/java/org/hisp/dhis/tracker/bundle/TrackedEntityDataValueAuditTest.java

```diff
@@ -107,30 +114,24 @@ void testTrackedEntityDataValueAuditCreate()
                 .setProgramStageInstances( List.of( psi ) )
                 .setAuditType( AuditType.DELETE ) );
 
-        assertNotNull( createdAudit );
-        assertNotNull( updatedAudit );
-        assertNotNull( deletedAudit );
-        assertFalse( createdAudit.isEmpty() );
-        assertFalse( updatedAudit.isEmpty() );
-        assertFalse( deletedAudit.isEmpty() );
-
-        createdAudit.forEach( a -> {
-            assertEquals( a.getAuditType(), AuditType.CREATE );
-            assertEquals( a.getDataElement().getUid(), dataElement.getUid() );
-            assertEquals( a.getProgramStageInstance().getUid(), psi.getUid() );
-        } );
-
-        updatedAudit.forEach( a -> {
-            assertEquals( a.getAuditType(), AuditType.UPDATE );
-            assertEquals( a.getDataElement().getUid(), dataElement.getUid() );
-            assertEquals( a.getProgramStageInstance().getUid(), psi.getUid() );
-            assertEquals( ORIGINAL_VALUE, a.getValue() );
-        } );
-
-        deletedAudit.forEach( a -> {
-            assertEquals( a.getAuditType(), AuditType.DELETE );
-            assertEquals( a.getDataElement().getUid(), dataElement.getUid() );
-            assertEquals( a.getProgramStageInstance().getUid(), psi.getUid() );
-        } );
+        assertAll( () -> assertNotNull( createdAudit ), () -> assertNotNull( updatedAudit ),
+            () -> assertNotNull( deletedAudit ) );
+        assertAuditCollection( createdAudit, AuditType.CREATE, ORIGINAL_VALUE );
+        assertAuditCollection( updatedAudit, AuditType.UPDATE, ORIGINAL_VALUE );
+        assertAuditCollection( deletedAudit, AuditType.DELETE, UPDATED_VALUE );
+
+    }
+
+    private void assertAuditCollection( List<TrackedEntityDataValueAudit> audits, AuditType auditType,
+        String expectedValue )
+    {
+        assertAll( () -> assertFalse( audits.isEmpty() ),
+            () -> assertEquals( auditType, audits.get( 0 ).getAuditType(),
```

Comment:
You might have misunderstood what I meant 😅

So assertions in integration tests should give us the context we need to potentially fix them with only that information and nothing else for example when the test only fails on Github.

You have duplicated the information we already get from `assertEquals` in a message supplier to it.

In your case, imagine `audits` has 2 or more elements and what you expect `expectedValue`/`auditType` is in the list, just not at position `0`. Unfortunately, this is happening regularly to us 😓 In such an error case with assertions structured like this you will only see an assertion message telling you that the `auditType` of element 0 is not what we expect. You will not know that maybe the element you are looking for is in the `audits` but just at position 2.

Try setting up a unit test with a list of audits and the assertion `assertAuditCollection` you wrote. Then add more than just one element, with the expected one in any position other than 0. This way you can iterate quickly and come up with assertions that give a useful error message. https://github.com/dhis2/dhis2-core/pull/11326 is another example of how adjusted some assertions using `assertAll`.